nx/lib/nx/binary_backend.ex

```diff
@@ -2033,6 +2051,29 @@ defmodule Nx.BinaryBackend do
     from_binary(t, new_data)
   end
 
+  @impl true
+  def triangular_solve(
+        %{type: output_type} = b_holder,
+        %{type: input_type, shape: {rows, rows}} = a,
+        b
+      ) do
+    a_matrix = a |> to_binary() |> binary_to_matrix(input_type, {rows, rows})
+    b_vector = b |> to_binary() |> binary_to_matrix(input_type, {0, 1}) |> List.flatten()
+
+    solution = List.duplicate(0, rows)
+
+    x =
+      Enum.with_index(a_matrix)
+      |> Enum.zip(b_vector)
+      |> Enum.reduce(solution, fn {{row, idx}, current_b}, previous_y ->
+        y = (current_b - dot_matrix(row, previous_y)) / Enum.at(row, idx)
+        replace_vector_element(previous_y, idx, y)
+      end)
```

Comment:
```suggestion
    x =
      Enum.with_index(a_matrix)
      |> Enum.zip(b_vector)
      |> Enum.reduce([], fn {{row, idx}, current_b}, previous_y ->
        y = (current_b - dot_matrix(row, previous_y)) / Enum.at(row, idx)
        previous_y ++ [y]
      end)
```

This works and avoids preallocation `solution` and replacing items in the list (which ends up being costly due to lists in elixir/erlang being linked-lists)

This only works because of the way `dot_matrix` is defined, dropping the remaining elements of the longer list (which for dot products is the same as dotting with a same-dimension matrix where the missing entries are 0)

Reply:
@polvalente This is great. Much more efficient. Thanks much!